geode-core/src/main/java/org/apache/geode/internal/cache/BucketAdvisor.java

```diff
@@ -2762,4 +2763,44 @@ void markShadowBucketAsDestroyed(String shadowBucketPath) {
   public boolean isShadowBucketDestroyed(String shadowBucketPath) {
     return destroyedShadowBuckets.getOrDefault(shadowBucketPath, false);
   }
+
+  boolean checkIfAllColocatedChildBucketsBecomePrimary() {
+    List<PartitionedRegion> colocatedChildPRs = getColocateNonShadowChildRegions();
+    if (colocatedChildPRs.size() > 0) {
+      for (PartitionedRegion partitionedRegion : colocatedChildPRs) {
+        Bucket bucket = partitionedRegion.getRegionAdvisor().getBucket(getBucket().getId());
+        if (bucket != null) {
+          BucketAdvisor bucketAdvisor = bucket.getBucketAdvisor();
+          if (!bucketAdvisor.checkIfAllColocatedChildBucketsBecomePrimary()) {
+            return false;
+          } else {
+            if (!checkIfBecomesPrimary()) {
+              return false;
+            }
+          }
+        }
+      }
+      return true;
+    }
+    return checkIfBecomesPrimary();
+  }
+
+  @NotNull
+  List<PartitionedRegion> getColocateNonShadowChildRegions() {
+    return ColocationHelper.getColocatedNonShadowChildRegions(pRegion);
+  }
+
+  private boolean checkIfBecomesPrimary() {
+    synchronized (this) {
+      try {
+        if (!isPrimary()) {
+          wait(10);
```

Comment:
Why the arbitrary wait 10ms?

Reply:
When we waiting for child buckets movePrimary to finish, we will allow sometime to see if it becomes primary. It can be argued how long it should wait.

Reply:
The getBucketAdvisor().checkIfAllColocatedChildBucketsBecomePrimary() will retry again, if it becomes primary in 11ms and will pass in the retry.
The reason added the pause is for a scenario that a child region is destroyed - while the leader is wait for bucketAdvisor state change, and never receive any if region is destroyed. However, the retry (after the pause) should detect this in the profile change and no longer wait for this child bucket when retrying.
I do think I may need to increase the pause to like 30 seconds or 60 seconds instead.